src/core/createCookie.js

```diff
@@ -23,58 +23,85 @@ const safeJSONParse = (object, cookieName) => {
   }
 };
 
+/**
+ * The purpose of this proxy is to cache the cookie so we don't have to
+ * read and deserialize it every time a piece of it is accessed.
+ */
+const createCookieProxy = (propertyID, cookieDomain = "") => {
+  const cookieName = `${ALLOY_COOKIE_NAME}_${propertyID}`;
+
+  let deserializedCookie;
+  let cookieHasBeenRead = false;
+
+  return {
+    get() {
+      // We don't read the cookie right when the cookie proxy is created
+      // because we don't know if the user has opted in. If the user
+      // hasn't opted in, we're legally obligated to not read the cookie.
+      // If a component tries to read something off the cookie though,
+      // we assume that the component has received word that the user
+      // has opted-in. The responsibility is on the component.
+      if (!cookieHasBeenRead) {
+        const serializedCookie = cookie.get(cookieName);
+        deserializedCookie =
+          serializedCookie && safeJSONParse(serializedCookie, cookieName);
+        cookieHasBeenRead = true;
+      }
+      return deserializedCookie;
+    },
+    set(updatedCookie) {
+      deserializedCookie = updatedCookie;
+      cookie.set(cookieName, updatedCookie, {
+        expires: ALLOY_COOKIE_TTL,
+        domain: cookieDomain || getTopLevelCookieDomain(window, cookie)
+      });
+    }
+  };
+};
+
 // TODO: Support passing a configurable expiry in the config when creating this cookie.
-const createCookie = (prefix, id, cookieDomain = "") => {
+const createCookie = (componentNamespace, propertyID, cookieDomain = "") => {
+  const cookieProxy = createCookieProxy(propertyID, cookieDomain);
+
   return {
     /**
      * Returns the value from AlloyCookie for a prefix and a key.
      * @param {...*} arg String key stored in alloy cookie under a component prefix.
      */
     get(name) {
-      const cookieName = `${ALLOY_COOKIE_NAME}_${id}`;
-      const currentCookie = cookie.get(cookieName);
-      const currentCookieParsed =
-        currentCookie && safeJSONParse(currentCookie, cookieName);
-
+      const currentCookie = cookieProxy.get();
       return (
-        currentCookieParsed &&
-        currentCookieParsed[prefix] &&
-        currentCookieParsed[prefix][name]
+        currentCookie &&
+        currentCookie[componentNamespace] &&
+        currentCookie[componentNamespace][name]
       );
     },
     /**
      * Updates the value  of a key from AlloyCookie for a prefix
      * @param {...*} arg Strings with key and value to be stored in alloy cookie.
      */
     set(key, value) {
-      const cookieName = `${ALLOY_COOKIE_NAME}_${id}`;
-      const currentCookie = cookie.get(cookieName)
-        ? safeJSONParse(cookie.get(cookieName))
-        : {};
-      const updatedCookie = {
-        ...currentCookie,
-        [prefix]: { ...currentCookie[prefix], [key]: value }
-      };
```

Comment:
Previously we were cloning the cookie object to prevent mutation of `currentCookie`, but we weren't doing the same thing in `remove()` when deleting and entry from the cookie object. For consistency, we either needed to do it in `remove()` as well, or not do it here. I thought that because any side effects of mutating the object were contained to this module, that it would be reasonable to just mutate the object. I open to disagreement here.